server/src/main/java/com/linecorp/centraldogma/server/internal/storage/ScheduledPurgingService.java

```diff
@@ -0,0 +1,252 @@
+                   .listPurged()
+                   .forEach(repoName -> cleanPurgedFile(repoName, "repository"));
+        });
+    }
+
+    @VisibleForTesting
+    void cleanPurgedFile(String name, String typeName) {
+        final File purged = new File(rootDir, name + SUFFIX_PURGED);
+        CompletableFuture.supplyAsync(() -> {
+            try {
+                Util.deleteFileTree(purged);
+            } catch (IOException e) {
+                logger.warn("Failed to delete a purged file " + typeName + ": {}", purged);
+            }
+            return null;
+        }, purgeWorker).join();
+    }
+
+    @VisibleForTesting
+    void schedulePurgeData(CommandExecutor commandExecutor,
+                           MetadataService metadataService) {
+        final long minAllowedTimestamp = System.currentTimeMillis() - maxRemovedRepositoryAgeMillis;
+        final Predicate<UserAndTimestamp> olderThanMinAllowed = timestamp -> {
+            if (timestamp == null) {
+                return false;
+            }
+            final Instant removedAt = Instant.from(
+                    DateTimeFormatter.ISO_INSTANT.parse(timestamp.timestamp()));
+            return removedAt.toEpochMilli() < minAllowedTimestamp;
+        };
+
+        purgeProject(commandExecutor, metadataService, olderThanMinAllowed);
+        purgeRepository(commandExecutor, metadataService, olderThanMinAllowed);
+    }
+
+    private void purgeProject(CommandExecutor commandExecutor,
+                              MetadataService metadataService,
+                              Predicate<UserAndTimestamp> olderThanMinAllowed) {
+        projectManager
+                .listRemoved()
+                .stream()
+                .map(project -> metadataService.getProject(project).join())
+                .forEach(metadata -> {
+                    if (olderThanMinAllowed.test(metadata.removal())) {
+                        commandExecutor
+                                .execute(Command.purgeProject(Author.SYSTEM, metadata.name())).join();
+                    }
+                });
+    }
+
+    private void purgeRepository(CommandExecutor commandExecutor,
+                                 MetadataService metadataService,
+                                 Predicate<UserAndTimestamp> olderThanMinAllowed) {
+        projectManager
+                .list()
+                .forEach((unused, project) -> project
+                        .repos()
+                        .listRemoved()
+                        .stream()
+                        .map(repoName -> metadataService.getRepo(project.name(), repoName).join())
+                        .forEach(metadata -> {
+                            if (olderThanMinAllowed.test(metadata.removal())) {
+                                commandExecutor.execute(
+                                        Command.purgeRepository(Author.SYSTEM, project.name(), metadata.name()))
+                                               .join();
```

Comment:
Could be indented this way for aesthetics:

```java
commandExecutor.execute(Command.purgeRepository(Author.SYSTEM,
                                                project.name(),
                                                metadata.name()))
               .join();
```

Reply:
I like aesthetics, too.